src/cirrus/lib2/events.py

```diff
@@ -0,0 +1,249 @@
+import json
+import os
+from contextlib import AbstractContextManager, contextmanager
+from datetime import datetime, timezone
+from logging import Logger, getLogger
+from typing import Dict
+
+import boto3
+
+from .enums import StateEnum, WFEventType
+from .eventdb import EventDB
+from .statedb import StateDB
+from .utils import SNSPublisher
+
+logger = getLogger(__name__)
+
+
+class WorkflowEventManager:
+    """A class for managing payload state change events, including storage of
+    state (DynamoDB), activity (Timestream), and notifications of decisions and/or status
+    changes (SNS)."""
+
+    def __init__(
+        self,
+        logger: Logger = logger,
+        boto3_session: boto3.Session = None,
+        statedb: StateDB = None,
+        eventdb: EventDB = None,
+        batch_size: int = 10,
+    ):
+        self.logger = logger
+        self._boto3_session = boto3_session if boto3_session else boto3.Session()
+        wf_event_topic_arn = os.getenv("CIRRUS_WORKFLOW_EVENT_TOPIC_ARN")
+        self.event_publisher = (
+            SNSPublisher(
+                wf_event_topic_arn,
+                logger=logger,
+                boto3_session=self._boto3_session,
+                batch_size=batch_size,
+            )
+            if wf_event_topic_arn
+            else None
+        )
+        self.statedb = (
+            statedb if statedb is not None else StateDB(session=self._boto3_session)
+        )
+        self.eventdb = (
+            eventdb if eventdb is not None else EventDB(session=self._boto3_session)
+        )
+
+    def flush(self):
+        if self.event_publisher:
+            self.event_publisher.execute()
+
+    @classmethod
+    @contextmanager
+    def handler(
```

Comment:
You could make `WorkflowEventManager` a context handler itself:

```
    def __enter__(self):
        return self

    def __exit__(self, et, ev, tb):
        self.flush()
```

Then to use it in a context you don't need to know/remember to call a special method, you just

```
with WorkflowEventManager() as wfm:
     ...
```

Reply:
Paired up and implemented!